keps/sig-node/3857-rro-mounts/README.md

```diff
@@ -0,0 +1,1080 @@
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+Support recursive read-only mounts for old runc and old kernel releases.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+A user wants to mount `/mnt`, includings its submounts such as `/mnt/usbstorage`, as read-only.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+Constraints: needs runc >= 1.1 && kernel >= 5.12.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+Risk: a manifest with `recursiveReadOnly: Enabled` will not work on old distros (See "Constraints" above).
+
+Mitigation: set `recursiveReadOnly: IfPossible`.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+
+### Core API
+Add `RecursiveReadOnly: (Disabled|IfPossible|Enabled)` to the [`VolumeMount`](https://github.com/kubernetes/kubernetes/blob/v1.26.1/pkg/apis/core/types.go#L1854-L1880) struct:
+- `(nil)`: keeps the legacy behavior
+- `Disabled`: makes the mountpoint non-recursively read-only
+- `IfPossible`: makes the mountpoint always read-only, but makes it recursively read-only only if possible.
+   Otherwise leaves it non-recursive, without raising an error.
+   This value is translated to `Enabled` or `Disabled` in the `VolumeMountStatus` struct (see the "Diff" below).
+- `Enabled`: makes the mountpoint recursively read-only if possible, otherwise raise an error (as long as the node components are aware of this feature).
+
+A pod manifest will look like this:
```

Comment:
do we need any validation that the Pod definition is not contradicts itself. Defining two mounts, one of them is subdirectory not marked as readonly, while `recursiveReadOnly` is set to true on other?

Reply:
No, the RRO flag only applies to the single mount struct where the flag belongs to.